client/src/components/categorical/category/index.js

```diff
@@ -101,138 +115,252 @@ class Category extends React.Component {
-    dispatch({
-      type: "categorical metadata filter none of these",
+  handleToggleAllClick = (categorySummary) => {
+    const isChecked = this.getSelectionState(categorySummary);
+    if (isChecked === "all") {
+      this.toggleNone(categorySummary);
+    } else {
+      this.toggleAll(categorySummary);
+    }
+  };
+
+  fetchAsyncProps = async (props) => {
+    const { annoMatrix, metadataField, colors } = props.watchProps;
+    const { crossfilter } = this.props;
+
+    const [categoryData, categorySummary, colorData] = await Category.fetchData(
+      annoMatrix,
       metadataField,
-      labels: categorySummary.categoryValues,
-    });
-    this.setState({ isChecked: false });
+      colors
+    );
+
+    return {
+      categoryData,
+      categorySummary,
+      colorData,
+      crossfilter,
+      ...this.updateColorTable(colorData),
+      handleCategoryToggleAllClick: () =>
+        this.handleToggleAllClick(categorySummary),
+    };
+  };
+
+  updateColorTable(colorData) {
+    // color table, which may be null
+    const { schema, colors, metadataField } = this.props;
+    const { colorAccessor, userColors, colorMode } = colors;
+    return {
+      isColorAccessor: colorAccessor === metadataField,
+      colorAccessor,
+      colorMode,
+      colorTable: createColorTable(
+        colorMode,
+        colorAccessor,
+        colorData,
+        schema,
+        userColors
+      ),
+    };
   }
 
-  toggleAll() {
+  toggleNone(categorySummary) {
     const { dispatch, metadataField } = this.props;
-    const { categorySummary } = this.state;
-    dispatch({
-      type: "categorical metadata filter all of these",
-      metadataField,
-      labels: categorySummary.categoryValues,
-    });
-    this.setState({ isChecked: true });
+    dispatch(
+      actions.selectCategoricalAllMetadataAction(
+        "categorical metadata filter none of these",
+        metadataField,
+        categorySummary.categoryValues,
+        false
+      )
+    );
   }
 
-  handleToggleAllClick() {
-    const { isChecked } = this.state;
-    if (isChecked) {
-      this.toggleNone();
-    } else {
-      this.toggleAll();
-    }
+  toggleAll(categorySummary) {
+    const { dispatch, metadataField } = this.props;
+    dispatch(
+      actions.selectCategoricalAllMetadataAction(
+        "categorical metadata filter all of these",
+        metadataField,
+        categorySummary.categoryValues,
+        true
+      )
+    );
   }
 
-  renderIsStillLoading() {
-    /*
-    We are still loading this category, so render a "busy" signal.
-    */
-    const { metadataField } = this.props;
+  render() {
+    const {
+      metadataField,
+      isExpanded,
+      categoricalSelection,
+      crossfilter,
+      colors,
+      annoMatrix,
+      isUserAnno,
+    } = this.props;
 
     const checkboxID = `category-select-${metadataField}`;
 
     return (
+      <CategoryCrossfilterContext.Provider value={crossfilter}>
```

Comment:
seems like the Provider gets the value from the Redux store connector above:

```ts
@connect((state, ownProps) => ({
  ...
  crossfilter: state.obsCrossfilter
}))
```

Can the consumer gets `crossfilter` from the Redux store too, so we don't need to use Context here?

E.g., 

```ts
// client/src/components/categorical/value/index.js:49

crossfilter: state.obsCrossfilter

...

render() {
  const { crossfilter } = this.props
  ...
}
```

Reply:
no, doing so would introduce races with the derived state.  The point of this context is to create a point-in-time snapshot of all state, for use by the child elements.  We could pass it down as a property, but given that it skips so many levels, I elected to keep the code cleaner by using a context.

Reply:
OH!! Because of `fetchAsyncProps()`?

I didn't dig into the details, but I'm imaging Redux store's `crossfilter` is used by `<Category />`, but Category uses `crossfilter` and its derived data asynchronously, so if `<Value />` uses the Redux store's `crossfilter` directly, `<Category />` and `<Value />` could be using different `crossfilter`?

Reply:
yes, the crossfilter could easily change while the promise is being resolved.   The async fetch of data, and the async (user-driven) crossfilter update will (and do) race.  You have to create consistent point-in-time data snapshots to render from.

Reply:
Makes sense now 👍 Thanks for the explanation!

CC: @colinmegill 😄